src/app/services/archive-project.service.ts

```diff
@@ -37,4 +38,74 @@ export class ArchiveProjectService {
   refreshProjects(): void {
     this.refreshProjectsEventSource.next();
   }
+
+  showArchiveProjectSuccessMessage(project: Project, archive: boolean): void {
+    this.snackBar
+      .open($localize`Successfully ${archive ? 'archived' : 'restored'} unit.`, $localize`Undo`)
+      .onAction()
+      .subscribe(() => {
+        this.undoArchiveProjectAction(project, archive ? 'unarchiveProject' : 'archiveProject');
+      });
+  }
+
+  undoArchiveProjectAction(project: Project, archiveFunctionName: string): void {
+    this[archiveFunctionName](project).subscribe({
+      next: (response: ArchiveProjectResponse) => {
+        this.updateProjectArchivedStatus(project, response.archived);
+        this.snackBar.open($localize`Action undone.`);
+      },
+      error: () => {
+        this.snackBar.open($localize`Error undoing action.`);
+      }
+    });
+  }
+
+  updateProjectArchivedStatus(project: Project, archived: boolean): void {
+    project.archived = archived;
+    if (archived) {
+      project.tags.push('archived');
+    } else {
+      project.tags.splice(project.tags.indexOf('archived'), 1);
+    }
```

Comment:
Should this be handled by a new ```archive()``` function that we define in ```Project```? Doing this would allow us to encapsulate the ```project.tags``` field.
```suggestion (pseudo-code)
    project.archive(archive);
```